### PR TITLE
feat(ui): Make UI errors recoverable. Fixes #3666

### LINKED ISSUE #3666: Malformed WorkflowTemplate crashes whole UI
Checklist:

 * [x] I've included the version.
 * [x] I've included reproduction steps.
 * [x] I've included the workflow YAML.
 * [x] I've included the logs.

**What happened**:
Deployed a WorkflowTemplate with kubectl (really helm) that was malformed, but still accepted by K8s API server (fit CRD apparently). This broke the WorkflowTemplate UI list view (http://localhost:2746/workflow-templates), see screenshot below.

argo cli would have caught it:
```
# argo template create -n argo src/test/resources/workflow-template.yaml
020/08/04 17:32:17 Failed to parse workflow template: error unmarshaling JSON: while decoding JSON: json: cannot unmarshal object into Go struct field Arguments.spec.templates.dag.tasks.arguments.parameters of type []v1alpha1.Parameter
```
but we will be deploying our workflow (-template) together with our application as a single helm chart because we want the declarative resource management, so argo cli isn't really an option.

**What you expected to happen**:
The WorkflowTemplate UI's list view should as least load and show me any WorkflowTemplates that are indeed syntactically correct. Put another way, I expect not to be prevented from working with my valid WorkflowTemplates, just because yesterday's newly deployed WorkflowTemplate is broken.

Particularly troubling is that the entire UI, including sidebar, disappears. Since it's only an ajax api request (`http://localhost:2746/api/v1/workflow-templates/argo`) that fails, I would expect to at least get a sensible error within the WorkflowTemplate UI about what has gone wrong.  Ideally I would even be allowed to see the valid WorkflowTemplates and be informed about the broken one separately, maybe with a warning icon on the list item. While this could presumably be fixed by making the api and frontend smarter, another 500 will crop up in the future and break the UI somewhere else, and I would love not to be kicked out of the UI entirely whenever the server fails to provide some data.

On that note, the error page seems to have a double-redirect which breaks the browser back button, so I can't even get back to the main page that I was on previously. *Should I post a separate issue about that?*

**How to reproduce it (as minimally and precisely as possible)**:
1. start with blank docker for mac kubernetes environment (haven't tried other setups, but I suspect any kubernetes environment should work)
2. setup argo
```
kubectl create ns argo
kubectl -n argo apply -f https://raw.githubusercontent.com/argoproj/argo/stable/manifests/quick-start-minimal.yaml
kubectl port-forward -n argo deploy/argo-server web
```
3. deploy a WorkflowTemplate that Kubernetes accepts (fits CRD spec), but is technically malformed in one place (maybe CRD spec isn't strict enough?)
```yaml
# workflow-template.yaml
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: some-template
spec:
  entrypoint: main
  templates:
    - name: main
      dag:
        tasks:
          - name: print
            template: print
            arguments:
              parameters:
                # this should be
                # parameters:
                #   name: someParam
                #   value: "Hello, world!"
                someParam: "Hello, world!"
    - name: print
      script:
        image:  busybox
        command: [sh]
        script: |
          echo input was: {{inputs.parameters.someParam}}
```
```
kubectl -n argo apply -f workflow-template.yaml
```
4. visit http://localhost:2746/workflow-templates
5. watch the entire ui disappear behind an unfriendly API 500 error page:

Screenshot 2020-08-04 at 17 29 54

**Anything else we need to know?**:

**Environment**:
Docker for Mac 2.3.0.4 (46911) with Kubernetes enabled

- Argo version:
```
argo: v2.9.4+20d2ace.dirty
  BuildDate: 2020-07-25T08:30:49Z
  GitCommit: 20d2ace3d5344db68ce1bc2a250bbb1ba9862613
  GitTreeState: dirty
  GitTag: v2.9.4
  GoVersion: go1.14.5
  Compiler: gc
  Platform: darwin/amd64
```
- Kubernetes version :
```
clientVersion:
  buildDate: "2019-12-13T11:51:44Z"
  compiler: gc
  gitCommit: 70132b0f130acc0bed193d9ba59dd186f0e634cf
  gitTreeState: clean
  gitVersion: v1.17.0
  goVersion: go1.13.4
  major: "1"
  minor: "17"
  platform: darwin/amd64
serverVersion:
  buildDate: "2020-01-15T08:18:29Z"
  compiler: gc
  gitCommit: e7f962ba86f4ce7033828210ca3556393c377bcc
  gitTreeState: clean
  gitVersion: v1.16.6-beta.0
  goVersion: go1.13.5
  major: "1"
  minor: 16+
  platform: linux/amd64
```

**Logs**

- server logs:

surprisingly, the server doesn't log the stacktrace seen in the UI.
```
time="2020-08-04T15:49:34Z" level=info authModes="[server client]" baseHRef=/ managedNamespace=argo namespace=argo secure=false
time="2020-08-04T15:49:34Z" level=warning msg="You are running in insecure mode. Learn how to enable transport layer security: https://github.com/argoproj/argo/blob/master/docs/tls.md"
time="2020-08-04T15:49:34Z" level=info msg="config map" name=workflow-controller-configmap
time="2020-08-04T15:49:34Z" level=info msg="SSO disabled"
time="2020-08-04T15:49:34Z" level=info msg="Starting Argo Server" version=v2.9.3
time="2020-08-04T15:49:34Z" level=info msg="Argo Server started successfully on http://localhost:2746"
```

- workflow-controller logs:

but the workflow-controller does (partly at least)
```
...
E0804 15:50:42.669594       1 reflector.go:125] pkg/mod/k8s.io/client-go@v0.0.0-20191225075139-73fd2ddc9180/tools/cache/reflector.go:98: Failed to list *v1alpha1.WorkflowTemplate: v1alpha1.WorkflowTemplateList.Items: v1alpha1.WorkflowTemplates: v1alpha1.WorkflowTemplate.Spec: v1alpha1.WorkflowTemplateSpec.WorkflowSpec: Templates: []v1alpha1.Template: v1alpha1.Template.DAG: v1alpha1.DAGTemplate.Tasks: []v1alpha1.DAGTask: v1alpha1.DAGTask.Arguments: v1alpha1.Arguments.Parameters: []v1alpha1.Parameter: decode slice: expect [ or n, but found {, error found in #10 byte of ...|ameters":{"someParam|..., bigger context ...|es":[{"dag":{"tasks":[{"arguments":{"parameters":{"someParam":"Hello, world!"}},"name":"print","temp|...
...
```

---

**Message from the maintainers**:

If you are impacted by this bug please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [ ] ~~I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.~~ N/A
* [ ] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

**I recommend the reviewer and I sit down and go over the changes.**

# Goals

The goal of this PR is to make the UI usable under error scenarios. Typically if an error occurs, we throw it and a global `<ErrorBoundary/>` render a generic "ops!" page. The user must then reload. This looses the user's context as well as any data entered into forms - very annoying.

Errors are now more localised in the interface. E.g. an error listing workflows just shows the error and keeps the page usable. The user can then change their search parameters. If the error was a 403, this would most useful.

# Changes

* Refactored handling of managed namespace in UI. Rather than have each component enforce this, it is done in `<App/>`. This simplifies many component.
* Add automatic redirection to the correct URLs for managed namespaces.
* Add `malformed-*.yaml` manifests to aid testing with malformed manifests.
* Removed `throw this.state.error` and replaced with `<ErrorNotice/>` so that the UI remains usable on error.
* `<ResourceEditor/>` now displays submission error in a way so that you can edit and retry.
* Simplified history by adding `saveHistory` functions and `this.url =`.

# Testing Notes

**I'm not expecting the reviewers to do this - this is for my notes**.

Does the UI stay usable?

* Try listing and deep-linking malformed resources. UI should still be usable.
* List each type of resource. 
* Creating each type of resource (both valid and malformed). Check namespace is correct.
* Deleting each type of resource. Check namespace is correct.
* Submitting each type of submittable resources. Check the correct namespace is used.

Can we deal with really bad errors?

* Run `yarn start` and turn off Argo Server so that you get gateway errors, try all pages.

Does login still work?

* Run in `make AUTH_MODE=client start` and verify that you are redirect to login page.

Does namespace preference still work?

* Run in `make NAMESPACED=false start` clear local storage and verify no redirection occurs (you should just list the at cluster scope).
* Change namespace in each page and verify that your local preference (stored in local storage) is updated.

# Known Issues

* The namespace preference does not affect URLs unless you reload the page. Fix to this is complex vs benefit IMHO.

### CODE REVIEW

File: ui/src/app/shared/components/resource-editor/resource-editor.tsx [hunk 1]
```diff
@@ -15,7 +15,7 @@ interface Props<T> {
     value: T;
     readonly?: boolean;
     editing?: boolean;
-    onSubmit?: (value: T) => void;
+    onSubmit?: (value: T) => Promise<any>;
```
Comment (alexec) [inline ui/src/app/shared/components/resource-editor/resource-editor.tsx:18]: small change, big impact

File: ui/src/app/shared/components/namespace-filter.tsx [hunk 2]
```diff
@@ -1,62 +1,4 @@
 import * as React from 'react';
-import {services} from '../services';
-import {ErrorPanel} from './error-panel';
 import {InputFilter} from './input-filter';
 
-interface Props {
```
Comment (alexec) [inline ui/src/app/shared/components/namespace-filter.tsx:6]: this component is now completely dumb

File: ui/src/app/shared/services/requests.ts [hunk 3]
```diff
@@ -39,28 +39,11 @@ export default {
         return auth(superagent.del(apiUrl(url)));
     },
 
-    loadEventSource(url: string, allowAutoRetry = false): Observable<string> {
+    loadEventSource(url: string): Observable<string> {
         return Observable.create((observer: Observer<any>) => {
             const eventSource = new EventSource(url);
-            let opened = false;
-            eventSource.onopen = () => {
```
Comment (alexec) [inline ui/src/app/shared/services/requests.ts:46]: this is canonical way to do this, I hope is more robust, certainly surfaces errors correctly

File: manifests/quick-start/sso/dex/dex-cm.yaml [hunk 4]
```diff
@@ -15,7 +15,7 @@ data:
     staticClients:
       - id: argo-server
         redirectURIs:
-          - https://localhost:2746/oauth2/callback
+          - http://localhost:2746/oauth2/callback
```
Comment (alexec) [inline manifests/quick-start/sso/dex/dex-cm.yaml:18]: fix

File: Procfile [hunk 5]
```diff
@@ -1,2 +1,2 @@
-controller: ALWAYS_OFFLOAD_NODE_STATUS=${ALWAYS_OFFLOAD_NODE_STATUS} OFFLOAD_NODE_STATUS_TTL=30s WORKFLOW_GC_PERIOD=30s UPPERIO_DB_DEBUG=1 ARCHIVED_WORKFLOW_GC_PERIOD=30s ./dist/workflow-controller --executor-image argoproj/argoexec:${VERSION} --namespaced --loglevel ${LOG_LEVEL}
-argo-server: UPPERIO_DB_DEBUG=1 ./dist/argo --loglevel ${LOG_LEVEL} server --namespaced --auth-mode ${AUTH_MODE} --secure=$SECURE
+controller: ALWAYS_OFFLOAD_NODE_STATUS=${ALWAYS_OFFLOAD_NODE_STATUS} OFFLOAD_NODE_STATUS_TTL=30s WORKFLOW_GC_PERIOD=30s UPPERIO_DB_DEBUG=1 ARCHIVED_WORKFLOW_GC_PERIOD=30s ./dist/workflow-controller --executor-image argoproj/argoexec:${VERSION} --namespaced=${NAMESPACED} --namespace ${NAMESPACE} --loglevel ${LOG_LEVEL}
```
Comment (alexec) [inline Procfile:1]: allow to run in non-namespaced mode in dev

File: ui/src/app/app.tsx [hunk 6]
```diff
@@ -110,7 +98,17 @@ export class App extends React.Component<{}, {version?: Version; popupProps: Pop
 
     public componentDidMount() {
         this.popupManager.popupProps.subscribe(popupProps => this.setState({popupProps}));
-        services.info.getVersion().then(version => this.setState({version}));
+        services.info
```
Comment (alexec) [inline ui/src/app/app.tsx:101]: 🔑  This is one of the key changes in this PR. This allows us to know what managed namespace we're running in almost immediately. We use this to create redirects that mean that child components no longer need to know about namespaces.

File: ui/src/app/app.tsx [hunk 7]
```diff
@@ -110,7 +98,17 @@ export class App extends React.Component<{}, {version?: Version; popupProps: Pop
...
+            .getVersion()
+            .then(version => this.setState({version}))
+            .then(() => services.info.getInfo())
+            .then(info => this.setState({namespace: info.managedNamespace || Utils.getCurrentNamespace() || ''}))
+            .catch(error => {
```
Comment (alexec) [inline ui/src/app/app.tsx:106]: I make an exception here - report this error in a popup. We still need to be able to get to the login page on 401.

File: ui/src/app/app.tsx [hunk 8]
```diff
@@ -124,24 +122,69 @@ export class App extends React.Component<{}, {version?: Version; popupProps: Pop
             <Provider value={providerContext}>
                 {this.state.popupProps && <Popup {...this.state.popupProps} />}
                 <Router history={history}>
-                    <Switch>
-                        <Redirect exact={true} path={uiUrl('')} to={workflowsUrl} />
-                        <Redirect from={timelineUrl} to={uiUrl('workflows')} />
+                    <Layout navItems={navItems} version={() => <>{this.state.version ? this.state.version.version : 'unknown'}</>}>
+                        <Notifications notifications={this.notificationsManager.notifications} />
+
                         <ErrorBoundary>
```
Comment (alexec) [inline ui/src/app/app.tsx:128]: move this inside the layout, so we can keep button even in case of a "ops!" error

File: ui/src/app/app.tsx [hunk 9]
```diff
@@ -124,24 +122,69 @@ export class App extends React.Component<{}, {version?: Version; popupProps: Pop
...
-                                })}
-                            </Layout>
+                            <Switch>
+                                <Route exact={true} strict={true} path={uiUrl('')}>
+                                    <Redirect to={workflowsUrl} />
+                                </Route>
+                                <Route exact={true} strict={true} path={timelineUrl}>
+                                    <Redirect to={workflowsUrl} />
+                                </Route>
+                                {this.state.namespace && (
```
Comment (alexec) [inline ui/src/app/app.tsx:136]: these are the redirects, it seems you should use `exact` and `strict` if you can

File: ui/src/app/archived-workflows/components/archived-workflow-list/archived-workflow-list.tsx [hunk 10]
```diff
@@ -20,9 +20,6 @@ import {ArchivedWorkflowFilters} from '../archived-workflow-filters/archived-wor
 
 interface State {
     pagination: Pagination;
-    loading: boolean;
```
Comment (alexec) [inline ui/src/app/archived-workflows/components/archived-workflow-list/archived-workflow-list.tsx:23]: we don't need these anymore, infer from other fields

File: ui/src/app/archived-workflows/components/archived-workflow-list/archived-workflow-list.tsx [hunk 11]
```diff
@@ -62,42 +56,32 @@ export class ArchivedWorkflowList extends BasePage<RouteComponentProps<any>, Sta
     }
 
     public render() {
-        if (this.state.loading) {
```
Comment (alexec) [inline ui/src/app/archived-workflows/components/archived-workflow-list/archived-workflow-list.tsx:65]: move error shims lower in the dom, so more the page remains responsive, you'll see this several times in this PR

File: ui/src/app/archived-workflows/components/archived-workflow-list/archived-workflow-list.tsx [hunk 12]
```diff
@@ -122,65 +106,61 @@ export class ArchivedWorkflowList extends BasePage<RouteComponentProps<any>, Sta
...
+            params.append('limit', this.state.pagination.limit.toString());
         }
-        const url = 'archived-workflows/' + namespace + '?' + params.toString();
-        history.pushState(null, '', uiUrl(url));
-        this.fetchArchivedWorkflows(namespace, selectedPhases, selectedLabels, minStartedAt, maxStartedAt, pagination);
+        return params;
+    }
+
+    private saveHistory() {
+        this.url = uiUrl('archived-workflows/' + this.state.namespace + '?' + this.filterParams.toString());
```
Comment (alexec) [inline ui/src/app/archived-workflows/components/archived-workflow-list/archived-workflow-list.tsx:132]: you'll see `saveHistory` several times as a new pattern for - saving history

File: ui/src/app/archived-workflows/components/archived-workflow-list/archived-workflow-list.tsx [hunk 13]
```diff
@@ -122,65 +106,61 @@ export class ArchivedWorkflowList extends BasePage<RouteComponentProps<any>, Sta
...
+                        pagination: {
+                            limit: pagination.limit,
+                            offset: pagination.offset,
+                            nextOffset: list.metadata.continue
+                        }
                     },
-                    loading: false
-                });
-                Utils.setCurrentNamespace(newNamespace);
+                    this.saveHistory
```
Comment (alexec) [inline ui/src/app/archived-workflows/components/archived-workflow-list/archived-workflow-list.tsx:154]: part of the `saveHistory` pattern is to invoke it after only on successful state update, again - you'll see this as a repeating pattern

File: ui/src/app/cluster-workflow-templates/components/cluster-workflow-template-details/cluster-workflow-template-details.tsx [hunk 14]
```diff
@@ -38,20 +40,15 @@ export class ClusterWorkflowTemplateDetails extends BasePage<RouteComponentProps
...
     public componentDidMount(): void {
-        services.info
-            .getInfo()
-            .then(info => this.setState({namespace: info.managedNamespace || 'default'}))
-            .catch(error => this.setState(error));
         services.clusterWorkflowTemplate
             .get(this.name)
             .then(template => this.setState({template}))
+            .then(() => services.info.getInfo())
+            .then(info => this.setState({namespace: info.managedNamespace || Utils.getCurrentNamespace() || 'default'}))
```
Comment (alexec) [inline ui/src/app/cluster-workflow-templates/components/cluster-workflow-template-details/cluster-workflow-template-details.tsx:47]: cluster templates an the one type where we cannot infer the namespace from the URL

File: ui/src/app/cluster-workflow-templates/components/cluster-workflow-template-summary-panel.tsx [hunk 15]
```diff
@@ -35,12 +34,9 @@ export const ClusterWorkflowTemplateSummaryPanel = (props: Props) => {
                         kind='ClusterWorkflowTemplate'
                         title='Update Cluster Workflow Template'
                         value={props.template}
-                        onSubmit={(value: WorkflowTemplate) => {
-                            services.clusterWorkflowTemplate
-                                .update(value, props.template.metadata.name)
-                                .then(clusterWorkflowTemplate => props.onChange(clusterWorkflowTemplate))
-                                .catch(err => props.onError(err));
-                        }}
+                        onSubmit={(value: WorkflowTemplate) =>
```
Comment (alexec) [inline ui/src/app/cluster-workflow-templates/components/cluster-workflow-template-summary-panel.tsx:37]: `onSubmit` now requires a promise which we display in the UI

File: ui/src/app/shared/components/error-notice.tsx [hunk 16]
```diff
@@ -1,13 +1,14 @@
...
 // Display an error notice.
 // If the error was a HTTP error (i.e. from super-agent), rather than just an unhelpful "Internal Server Error",
 // it will display any message in the body.
-export const ErrorNotice = (props: {error: Error & {response?: {body: {message?: string}}}}) => (
-    <Notice>
-        <i className='fa fa-exclamation-triangle status-icon--failed' />
-        {props.error.message}
+export const ErrorNotice = (props: {style?: CSSProperties; error: Error & {response?: {body: {message?: string}}}}) => (
+    <Notice style={props.style}>
+        <PhaseIcon value='Error' /> {props.error.message || 'Unknown error. Open your browser error console for more information.'}
```
Comment (alexec) [inline ui/src/app/shared/components/error-notice.tsx:11]: there are rare cases where we don't get info in the error - ask the user to look in the console

File: ui/src/app/archived-workflows/components/archived-workflow-details/archived-workflow-details.tsx [hunk 17]
```diff
@@ -135,6 +133,9 @@ export class ArchivedWorkflowDetails extends BasePage<RouteComponentProps<any>,
     }
 
     private renderArchivedWorkflowDetails() {
+        if (this.state.error) {
```
Comment (alexec) [inline ui/src/app/archived-workflows/components/archived-workflow-details/archived-workflow-details.tsx:136]: Screenshot 2020-08-05 at 10 58 13

File: ui/src/app/archived-workflows/components/archived-workflow-details/archived-workflow-details.tsx [hunk 18]
```diff
@@ -212,12 +213,11 @@ export class ArchivedWorkflowDetails extends BasePage<RouteComponentProps<any>,
                                 },
                                 spec: this.state.workflow.spec
                             }}
-                            onSubmit={(value: Workflow) => {
+                            onSubmit={(value: Workflow) =>
```
Comment (alexec) [inline ui/src/app/archived-workflows/components/archived-workflow-details/archived-workflow-details.tsx:216]: Screenshot 2020-08-05 at 10 59 04

File: ui/src/app/archived-workflows/components/archived-workflow-list/archived-workflow-list.tsx [hunk 19]
```diff
@@ -122,65 +106,61 @@ export class ArchivedWorkflowList extends BasePage<RouteComponentProps<any>, Sta
...
+                );
             })
-            .catch(error => this.setState({error, loading: false}));
+            .catch(error => this.setState({error}));
     }
 
     private renderWorkflows() {
+        if (this.state.error) {
```
Comment (alexec) [inline ui/src/app/archived-workflows/components/archived-workflow-list/archived-workflow-list.tsx:161]: Screenshot 2020-08-05 at 11 00 33

File: ui/src/app/cluster-workflow-templates/components/cluster-workflow-template-details/cluster-workflow-template-details.tsx [hunk 20]
```diff
@@ -102,10 +99,13 @@ export class ClusterWorkflowTemplateDetails extends BasePage<RouteComponentProps
     }
 
     private renderClusterWorkflowTemplate() {
+        if (this.state.error) {
```
Comment (alexec) [inline ui/src/app/cluster-workflow-templates/components/cluster-workflow-template-details/cluster-workflow-template-details.tsx:102]: Screenshot 2020-08-05 at 11 00 33

File: ui/src/app/cluster-workflow-templates/components/cluster-workflow-template-summary-panel.tsx (hunk 15 above)
Comment (alexec) [reply ui/src/app/cluster-workflow-templates/components/cluster-workflow-template-summary-panel.tsx:37]: Screenshot 2020-08-05 at 11 07 21

File: ui/src/app/cron-workflows/components/cron-workflow-details/cron-workflow-details.tsx [hunk 21]
```diff
@@ -90,12 +88,13 @@ export class CronWorkflowDetails extends BasePage<RouteComponentProps<any>, Stat
     }
 
     private renderCronWorkflow() {
+        if (this.state.error) {
```
Comment (alexec) [inline ui/src/app/cron-workflows/components/cron-workflow-details/cron-workflow-details.tsx:91]: Screenshot 2020-08-05 at 11 07 52

File: ui/src/app/cron-workflows/components/cron-workflow-list/cron-workflow-list.tsx [hunk 22]
```diff
@@ -100,20 +91,22 @@ export class CronWorkflowList extends BasePage<RouteComponentProps<any>, State>
...
+
+    private fetchCronWorkflows(namespace: string): void {
+        services.cronWorkflows
+            .list(namespace)
+            .then(cronWorkflows => this.setState({namespace, cronWorkflows}, this.saveHistory))
+            .catch(error => this.setState({error}));
     }
 
     private renderCronWorkflows() {
+        if (this.state.error) {
```
Comment (alexec) [inline ui/src/app/cron-workflows/components/cron-workflow-list/cron-workflow-list.tsx:107]: Screenshot 2020-08-05 at 11 08 24

File: ui/src/app/cron-workflows/components/cron-workflow-summary-panel.tsx [hunk 23]
```diff
@@ -72,12 +70,11 @@ export const CronWorkflowSummaryPanel = (props: Props) => {
                             // magic - we get the latest from the server and then apply the changes from the rendered version to this
                             const original = props.cronWorkflow;
                             const patch = jsonMergePatch.generate(original, value) || {};
-                            services.cronWorkflows
+                            return services.cronWorkflows
```
Comment (alexec) [inline ui/src/app/cron-workflows/components/cron-workflow-summary-panel.tsx:73]: Screenshot 2020-08-05 at 11 09 31

File: ui/src/app/userinfo/components/user-info.tsx [hunk 24]
```diff
@@ -25,12 +26,10 @@ export class UserInfo extends BasePage<RouteComponentProps<any>, State> {
     }
 
     public render() {
-        if (this.state.error) {
-            throw this.state.error;
-        }
         return (
             <Page title='User Info' toolbar={{breadcrumbs: [{title: 'User Info'}]}}>
                 <div className='argo-container'>
+                    {this.state.error && <ErrorNotice error={this.state.error} />}
```
Comment (alexec) [inline ui/src/app/userinfo/components/user-info.tsx:32]: Screenshot 2020-08-05 at 11 10 25

File: ui/src/app/workflow-templates/components/workflow-template-details/workflow-template-details.tsx [hunk 25]
```diff
@@ -101,10 +99,13 @@ export class WorkflowTemplateDetails extends BasePage<RouteComponentProps<any>,
     }
 
     private renderWorkflowTemplate() {
+        if (this.state.error) {
```
Comment (alexec) [inline ui/src/app/workflow-templates/components/workflow-template-details/workflow-template-details.tsx:102]: Screenshot 2020-08-05 at 11 10 59

File: ui/src/app/workflow-templates/components/workflow-template-list/workflow-template-list.tsx [hunk 26]
```diff
@@ -99,20 +89,22 @@ export class WorkflowTemplateList extends BasePage<RouteComponentProps<any>, Sta
...
+
+    private fetchWorkflowTemplates(namespace: string): void {
+        services.workflowTemplate
+            .list(namespace)
+            .then(templates => this.setState({templates}, this.saveHistory))
+            .catch(error => this.setState({error}));
     }
 
     private renderTemplates() {
+        if (this.state.error) {
```
Comment (alexec) [inline ui/src/app/workflow-templates/components/workflow-template-list/workflow-template-list.tsx:105]: Screenshot 2020-08-05 at 11 11 40

File: ui/src/app/workflow-templates/components/workflow-template-summary-panel.tsx [hunk 27]
```diff
@@ -34,12 +33,11 @@ export const WorkflowTemplateSummaryPanel = (props: Props) => {
                 <div className='white-box__details'>
                     <ResourceEditor
                         value={props.template}
-                        onSubmit={(value: WorkflowTemplate) => {
+                        onSubmit={(value: WorkflowTemplate) =>
```
Comment (alexec) [inline ui/src/app/workflow-templates/components/workflow-template-summary-panel.tsx:36]: Screenshot 2020-08-05 at 11 13 17

File: ui/src/app/workflows/components/workflows-list/workflows-list.tsx [hunk 28]
```diff
@@ -290,10 +282,12 @@ export class WorkflowsList extends BasePage<RouteComponentProps<any>, State> {
     }
 
     private renderWorkflows() {
+        if (this.state.error) {
```
Comment (alexec) [inline ui/src/app/workflows/components/workflows-list/workflows-list.tsx:285]: Screenshot 2020-08-05 at 11 15 25

**Comment by alexec** [comment]
@jessesuen you asked about whether or not having malformed resources (i.e. kubectl applied) would impact the controller. I loaded the malformed resources (1x cron workflow, 1x template template, 1x cluster workflow template, 1x workflow) and I could still create workflows directly.

Problems:

* I created another valid cron workflow - this did not schedule. 
* I could not create a workflow from a valid workflow template
* I could not create a workflow from a valid cluster workflow template.

These resolved after deleting the resource - I did not need to restart the container.

File: ui/src/app/app.tsx (hunk 6 above)
Comment (simster7) [reply ui/src/app/app.tsx:101]: Great change

**Pull request merged**